group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -747,6 +798,11 @@ private void throwIfMemberEpochIsInvalid(
         List<ConsumerGroupHeartbeatRequestData.TopicPartitions> ownedTopicPartitions
     ) {
         if (receivedMemberEpoch > member.memberEpoch()) {
+            // If a static member rejoins, it's previous epoch would be -2. In such a
+            // case, we don't need to fence the member.
+            if (member.memberEpoch() == LEAVE_GROUP_STATIC_MEMBER_EPOCH && receivedMemberEpoch == 0) {
+                return;
+            }
```

Comment:
I don't fully get this one. Could you please elaborate?

Reply:
This is mainly needed for the static member rejoin case. Let's say a static member with instance id `id` departed. When it departs, we would write a member epoch value of -2 against it. Now, if a new static member joins with the same instance id `id` and a member epoch value of 0, then without this condition, the rejoin would always fail with `FencedMemberEpochException`. This condition was added to avoid the same.

Reply:
I see. Would it make sense to put this condition first in the method, before `if (receivedMemberEpoch > member.memberEpoch())`? I got confused by the fact that it is within the if branch.

Reply:
Yeah, that makes sense as well. I placed it inside the if condition because this condition at hand shows up only when received epoch > current member epoch. But it should be ok to have it outside as you said. I have made the change.